yash-builtin/src/read/assigning.rs

```diff
@@ -0,0 +1,419 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! Assigning the input to variables
+
+use yash_env::semantics::Field;
+use yash_env::variable::Scope;
+use yash_env::variable::Value;
+use yash_env::variable::Variable;
+use yash_env::Env;
+use yash_semantics::expansion::attr::AttrChar;
+use yash_semantics::expansion::attr_strip::Strip as _;
+use yash_semantics::expansion::quote_removal::skip_quotes;
+use yash_semantics::expansion::split::Class;
+use yash_semantics::expansion::split::Ifs;
+use yash_syntax::source::pretty::Message;
+use yash_syntax::source::pretty::MessageBase as _;
+
+pub use crate::typeset::AssignReadOnlyError as Error;
+
+/// Converts errors to a message.
+///
+/// Returns `None` if `errors` is empty.
+pub fn to_message(errors: &[Error]) -> Option<Message> {
+    let mut message = Message::from(errors.first()?);
+    let other_errors = errors[1..].iter().map(Error::main_annotation);
+    message.annotations.extend(other_errors);
+    Some(message)
+}
+
+/// Assigns the text to variables.
+///
+/// This function performs field splitting on the text and assigns the resulting
+/// fields to the variables. When there are more fields than variables, the last
+/// variable receives all remaining fields, including the field separators, but
+/// not trailing whitespace separators. When there are fewer fields than
+/// variables, the remaining variables are set to empty strings.
+///
+/// The return value is a vector of errors that occurred while assigning the
+/// variables. The vector is empty if no error occurred.
+pub fn assign(
+    env: &mut Env,
+    text: &[AttrChar],
+    variables: Vec<Field>,
+    last_variable: Field,
+) -> Vec<Error> {
+    #[rustfmt::skip]
+    let ifs = match env.variables.get("IFS") {
+        Some(&Variable { value: Some(Value::Scalar(ref value)), ..  }) => value,
+        // TODO If the variable is an array, should we ignore it?
+        _ => Ifs::DEFAULT,
+    };
+    let ifs = ifs.to_owned();
+    let ifs = Ifs::new(&ifs);
+
+    let mut ranges = ifs.ranges(text.iter().copied());
+
+    // Assign variables but the last
+    let mut errors = variables
+        .into_iter()
+        .filter_map(|var_name| {
+            let value = ranges.next().map(|r| &text[r]).unwrap_or_default();
+            assign_one(env, var_name, value).err()
+        })
+        .collect::<Vec<_>>();
+
+    // Assign the last
+    let range = match ranges.next() {
+        None => 0..0,
+        Some(range) => match ranges.next() {
+            None => range,
+            Some(_range) => {
+                let end = text
+                    .iter()
+                    .rposition(|&c| ifs.classify_attr(c) != Class::IfsWhitespace)
+                    .unwrap()
+                    + 1;
+                range.start..end
```

Comment:
The `rposition` returns the position of the last non-IFS whitespace, which should be included in the result, hence the `+ 1`. What is exactly a problem with that?

Reply:
I think you're messing up yourself.
Since `rposition` finds the last non-IFS whitespace, all characters thereafter are IFS whitespace, right? Those IFS whitespace characters are not included in the result because they just appear after the computed end position of the result range.